mods/ra/maps/allies-09a/allies09a.lua

```diff
@@ -0,0 +1,128 @@
+--[[
+   Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+   This file is part of OpenRA, which is free software. It is made
+   available to you under the terms of the GNU General Public License
+   as published by the Free Software Foundation, either version 3 of
+   the License, or (at your option) any later version. For more
+   information, see COPYING.
+]]
+
+
+lstReinforcements =
+{
+	actors = { "mcv"},
+	entryPath = { AlliedMCVEntry.Location, Unload1.Location },
+	exitPath = { AlliedMCVEntry.Location }
+}
+
+ExtractionHelicopterType = "tran"
+ExtractionPath = { HeliWP01.Location, HeliWP02.Location, HeliWP03.Location }
+
+kosyginType = "gnrl"
+contectKosygin = false
+
+
+
+MissionAccomplished = function()
+	Media.PlaySpeechNotification(player, "MissionAccomplished")
+end
+
+MissionFailed = function()
+	Media.PlaySpeechNotification(player, "MissionFailed")
+end
+
+InitialAlliedReinforcements = function()
+	Trigger.AfterDelay(DateTime.Seconds(1), function()
+		Media.PlaySpeechNotification(player, "ReinforcementsArrived")
+		Reinforcements.ReinforceWithTransport(player, "lst.reinforcement", lstReinforcements.actors, lstReinforcements.entryPath, lstReinforcements.exitPath)
+	end)
+end
+
+
+RescueFailed = function()
+	Media.PlaySpeechNotification(player, "ObjectiveNotMet")
+	player.MarkFailedObjective(KosyginSurviveObjective)
+end
+
+CreateKosygin = function()
+	player.MarkCompletedObjective(UseSpyObjective)
+	Media.PlaySpeechNotification(player, "ObjectiveMet")
+	kosygin = Actor.Create(kosyginType, true, { Location = KosyginSpawnPoint.Location, Owner = player })
+	Trigger.OnKilled(kosygin, RescueFailed)
+	ExtractObjective = player.AddPrimaryObjective("Extract Kosygin.")
+	Trigger.AfterDelay(DateTime.Seconds(1), function() Media.PlaySpeechNotification(player, "TargetFreed") end)
+end
+
+
+InfiltrateForwardCenter = function()
+	Trigger.OnInfiltrated(USSRFC, function()
+		if contectKosygin == false then
+			contectKosygin = true
+			CreateKosygin()
+		end
+	end)
+
+	Trigger.OnKilledOrCaptured(USSRFC, function()
+		if not player.IsObjectiveCompleted(UseSpyObjective) then
+			player.MarkFailedObjective(UseSpyObjective)
+		end
+	end)
+end
+
+
+Tick = function()
+	ussr.Cash = 5000
+end
```

Comment:
We'll want to check for `player.HasNoRequiredUnits`.

Reply:
Add "HasNoRequiredUnits"